std-bits/table/src/main/java/org/enso/table/read/DelimitedReader.java

```diff
@@ -0,0 +1,339 @@
+      long skipRows,
+      long rowLimit,
+      int maxColumns,
+      boolean keepInvalidRows,
+      boolean warningsAsErrors) {
+    if (delimiter.isEmpty()) {
+      throw new IllegalArgumentException("Empty delimiters are not supported.");
+    }
+    if (delimiter.length() > 1) {
+      throw new IllegalArgumentException(
+          "Delimiters consisting of multiple characters or code units are not supported.");
+    }
+
+    this.delimiter = delimiter.charAt(0);
+
+    if (quote != null) {
+      if (quote.isEmpty()) {
+        throw new IllegalArgumentException(
+            "Empty quotes are not supported. Set the quote to `Nothing` to disable quotes.");
+      }
+      if (quote.length() > 1) {
+        throw new IllegalArgumentException(
+            "Quotes consisting of multiple characters or code units are not supported.");
+      }
+
+      quoteCharacter = quote.charAt(0);
+      if (quoteCharacter == noQuoteCharacter) {
+        throw new IllegalArgumentException("Illegal quote character.");
+      }
+    } else {
+      quoteCharacter = noQuoteCharacter;
+    }
+
+    if (quoteEscape != null) {
+      if (quoteEscape.isEmpty()) {
+        throw new IllegalArgumentException(
+            "Empty quote escapes are not supported. Set the escape to `Nothing` to disable escaping quotes.");
+      }
+      if (quoteEscape.length() > 1) {
+        throw new IllegalArgumentException(
+            "Quote escapes consisting of multiple characters or code units are not supported.");
+      }
+
+      quoteEscapeCharacter = quoteEscape.charAt(0);
+    } else {
+      quoteEscapeCharacter = noQuoteCharacter;
+    }
+
+    this.headerBehavior = headerBehavior;
+    this.skipRows = skipRows;
+    this.rowLimit = rowLimit;
+    this.maxColumns = maxColumns;
+    this.keepInvalidRows = keepInvalidRows;
+    this.warningsAsErrors = warningsAsErrors;
+
+    parser = setupCsvParser(inputStream);
+  }
+
+  /** Creates a {@code CsvParser} according to the settings specified at construction. */
+  private CsvParser setupCsvParser(InputStream inputStream) {
+    CsvParserSettings settings = new CsvParserSettings();
+    settings.setHeaderExtractionEnabled(false);
+    CsvFormat format = new CsvFormat();
+    format.setDelimiter(delimiter);
+    format.setQuote(quoteCharacter);
+    format.setQuoteEscape(quoteEscapeCharacter);
+    settings.setFormat(format);
+    settings.setMaxCharsPerColumn(-1);
+    settings.setMaxColumns(maxColumns);
+    settings.setSkipEmptyLines(false);
+    settings.setKeepQuotes(true);
+    CsvParser parser = new CsvParser(settings);
+    parser.beginParsing(inputStream);
+    return parser;
+  }
+
+  /**
+   * Parses a cell.
+   *
+   * <p>Currently just handles quotes. In the future it will be extended to delegate to a parsing
+   * strategy for values.
+   */
+  private String parseCell(String cell) {
+    if (cell == null) return null;
+
+    // TODO [RW] here we can plug-in some more complex parsing logic, to be done
```

Comment:
not sure parsing will be here - feels like might be in the column builders

Reply:
Hmmm, I guess you are right.

Well I can remove the TODO, will be able to figure out where to put parser without this comment :D